### PR TITLE
[WIP] [RFC] Fixing mypy errors produced by numpy 1.20

### PR DESCRIPTION
**Pull request by 0x41head**
## Motivation
#2293 

## Description of the changes
Feel free to use this as a starting point for fixing more errors

### CODE REVIEW

**Comment by Devanshu24** [comment]
Thanks @0x41head ! Let us split the files so that we don't end up doing the same thing

Maybe since you have already done some changes to the files, we could do it this way (for now, we could change it if need be)

You could do the `/importance`, `/multi_objective`, `/pruners`, `/samplers`
I will start with, `/integration`, `/samplers_tests`, `/multi_objective_tests`, `/integration_tests`

Sounds good?
___
I used these for reference
- [ ] optuna/importance/_fanova/_fanova.py
- [ ]  optuna/importance/_fanova/_tree.py
- [ ]  optuna/integration/botorch.py
- [ ]  optuna/integration/cma.py
- [ ]  optuna/integration/sklearn.py
- [ ]  optuna/multi_objective/_hypervolume/wfg.py
- [ ]  optuna/multi_objective/samplers/_motpe.py
- [ ]  optuna/pruners/_percentile.py
- [ ]  optuna/samplers/_tpe/multi_objective_sampler.py
- [ ]  optuna/samplers/_tpe/multivariate_parzen_estimator.py
- [ ]  optuna/samplers/_tpe/parzen_estimator.py
- [ ]  optuna/samplers/_tpe/sampler.py
- [ ]  tests/importance_tests/fanova_tests/test_tree.py
- [ ]  tests/integration_tests/test_chainer.py
- [ ]  tests/integration_tests/test_sklearn.py
- [ ]  tests/multi_objective_tests/hypervolume_tests/test_wfg.py
- [ ]  tests/multi_objective_tests/samplers_tests/test_motpe.py
- [ ]  tests/samplers_tests/test_grid.py
- [ ]  tests/samplers_tests/tpe_tests/test_multi_objective_sampler.py
- [ ]  tests/samplers_tests/tpe_tests/test_parzen_estimator.py
- [ ]  tests/storages_tests/rdb_tests/test_with_server.py

**Comment by hvy** [comment]
Thanks for the quick action. If other's can push to your branch, maybe we can work on a single PR. We could also ignore (`# type: ignore`) remaining ones and have this PR merge first but this would make it a bit of a hassle to make sure that we don't forget any. Open for ideas how we should continue from here. 

Could you by the way go through the CI failures? Some are purely cosmetic (`black`) while some seem to be more severe and related to the logical changes.

**Comment by 0x41head** [comment]
Seems fine to me @Devanshu24 . Sure, I'll look into it @hvy

**Comment by HideakiImamura** [comment]
The tests are failing now. Do you pass unit-tests locally?

**Comment by 0x41head** [comment]
Yes, I was aware @HideakiImamura . I just wanted to let @Devanshu24  know about this.

**Comment by 0x41head** [comment]
@HideakiImamura are the rdb storage tests from circleCI still relevant or have the become outdated ? Asking since they seem to be removed from the docs.

**Comment by 0x41head** [comment]
Sorry for the inactivity on this PR. I think the latest commit solves this issue. It would be really helpful if a contributor/reviewer could confirm this for me as well.

File: optuna/importance/_fanova/_tree.py [hunk 1]
```diff
@@ -136,8 +136,8 @@ def _get_marginalized_statistics(self, feature_vector: numpy.ndarray) -> Tuple[f
             node_indices.append(node_index)
             active_features_cardinalities.append(_get_cardinality(search_spaces))
 
-        node_indices = numpy.array(node_indices, dtype=numpy.int32)
-        active_features_cardinalities = numpy.array(active_features_cardinalities)
+        node_indices = numpy.array(node_indices, dtype=numpy.int32).tolist()
+        active_features_cardinalities = numpy.array(active_features_cardinalities).tolist()
```
Comment (hvy) [inline optuna/importance/_fanova/_tree.py:140]: Seems like unnecessary conversions are taking place here. Should we give them new variable names or just skip these lines instead?

File: optuna/visualization/_contour.py [hunk 2]
```diff
@@ -302,7 +302,7 @@ def _generate_contour_subplot(
     # If contours_coloring='heatmap' is specified, reversescale argument of go.Contour does not
     # work correctly. See https://github.com/pfnet/optuna/issues/606.
     colorscale = plotly.colors.PLOTLY_SCALES["Blues"]
-    if direction == StudyDirection.MAXIMIZE:
+    if direction == StudyDirection.MINIMIZE:
```
Comment (hvy) [inline optuna/visualization/_contour.py:305]: Can we omit this unrelated change from this PR?

Comment (0x41head) [reply optuna/visualization/_contour.py:305]: Sorry about that.

**Comment by HideakiImamura** [comment]
@0x41head In my local environment with `numpy==1.20.1`, the `mypy .` command fails. The log is [here](https://gist.github.com/HideakiImamura/26f18e46254a3be6ba52409aeb6a1b3e). The version of `mypy` is `0.790`. Do you succeed in your local environment?

**Comment by 0x41head** [comment]
@HideakiImamura After re-cloning the repo in my local environment. I am running into the same errors as you. Thanks for bringing it to my attention, I'll start working on it ASAP.

**Comment by 0x41head** [comment]
I think my latest commit takes care of this issue. @HideakiImamura ( or anyone else ) can you please confirm this in your local environment .

File: optuna/integration/sklearn.py [hunk 3]
```diff
@@ -195,7 +195,7 @@ def __init__(
         y: Optional[Union[OneDimArrayLikeType, TwoDimArrayLikeType]],
         cv: "BaseCrossValidator",
         enable_pruning: bool,
-        error_score: Union[Number, str],
+        error_score: Union[Number, float, str],
```
Comment (hvy) [inline optuna/integration/sklearn.py:198]: So `float` isn't a `Number`?

File: optuna/integration/sklearn.py [hunk 4]
```diff
@@ -666,7 +668,7 @@ def __init__(
         param_distributions: Mapping[str, distributions.BaseDistribution],
         cv: Optional[Union["BaseCrossValidator", int]] = 5,
         enable_pruning: bool = False,
-        error_score: Union[Number, str] = np.nan,
+        error_score: Union[Number, float, str] = np.nan,
```
Comment (hvy) [inline optuna/integration/sklearn.py:671]: `isinstance(np.nan, Number)` should be `True` but is this change still required?

File: optuna/multi_objective/_hypervolume/wfg.py [hunk 5]
```diff
@@ -15,14 +15,14 @@ class WFG(BaseHypervolume):
     """
 
     def __init__(self) -> None:
-        self._reference_point = None
```
Comment (hvy) [inline optuna/multi_objective/_hypervolume/wfg.py:18]: I think it's clearer to type hint this member as an `Optional[np.array]` instead of changing the logic (and add `assert self._reference_point is not None` where needed).

File: optuna/samplers/_tpe/multivariate_parzen_estimator.py [hunk 6]
```diff
@@ -72,7 +72,7 @@ def __init__(
                 categorical_weights = self._calculate_categorical_params(observations, param_name)
             else:
                 mus, sigmas = self._calculate_numerical_params(observations, param_name)
-                categorical_weights = None
+                categorical_weights = np.array([None])
```
Comment (hvy) [inline optuna/samplers/_tpe/multivariate_parzen_estimator.py:75]: Ditto.

File: optuna/samplers/_tpe/multivariate_parzen_estimator.py [hunk 7]
```diff
@@ -259,26 +259,29 @@ def _transform_from_uniform(
 
             assert isinstance(distribution, _DISTRIBUTION_CLASSES)
             if isinstance(distribution, distributions.UniformDistribution):
-                transformed[param_name] = samples
+                transformed[param_name] = np.array(samples)
```
Comment (hvy) [inline optuna/samplers/_tpe/multivariate_parzen_estimator.py:262]: Isn't `samples` always an ndarray here given the type hints from the method signature?

File: optuna/samplers/_tpe/parzen_estimator.py [hunk 8]
```diff
@@ -80,17 +80,17 @@ def _calculate(
                 order: List[int] = []
             else:  # When mus.size is greater than 0.
                 # We decide the place of the  prior.
-                order = numpy.argsort(mus).astype(int)
+                order = numpy.argsort(mus).astype(int).tolist()
```
Comment (hvy) [inline optuna/samplers/_tpe/parzen_estimator.py:83]: This might be a detail but again, this'll incur an unnecessary copy. How about always making `order` an ndarray instead?

File: tests/importance_tests/fanova_tests/test_tree.py [hunk 9]
```diff
@@ -207,7 +206,9 @@ def test_tree_split_sizes(tree: _FanovaTree, node_index: NodeIndex, expected: Li
 def test_tree_subtree_active_features(
     tree: _FanovaTree, node_index: NodeIndex, expected: List[bool]
 ) -> None:
-    active_features: numpy.ndarray = tree._subtree_active_features[node_index] == expected
+    active_features: numpy.ndarray = numpy.array(
+        [tree._subtree_active_features[node_index] == expected], dtype=bool
+    )
```
Comment (hvy) [inline tests/importance_tests/fanova_tests/test_tree.py:211]: I think that the logic here used to be correct. It's the signature of `_precompute_subtree_active_features` that needs to be fixed.

**Review by hvy — COMMENTED** [review]
Left some additionally comments. You probably know better regarding some of them but please let me know.

File: optuna/multi_objective/_hypervolume/wfg.py [hunk 10]
```diff
@@ -48,11 +50,13 @@ def _compute_rec(self, solution_set: np.ndarray) -> float:
         return volume
 
     def _compute_exclusive_hv(self, point: np.ndarray, solution_set: np.ndarray) -> float:
-        volume = _compute_2points_volume(point, self._reference_point)
+        volume = _compute_2points_volume(point, np.array(self._reference_point))
```
Comment (hvy) [inline optuna/multi_objective/_hypervolume/wfg.py:53]: Let's avoid copies. Or better, you could add an assertion that it's not `None` prior to this line similar to `_compute_rec`.
```suggestion
        volume = _compute_2points_volume(point, np.asarray(self._reference_point))
```

File: optuna/multi_objective/_hypervolume/wfg.py [hunk 11]
```diff
@@ -48,11 +50,13 @@ def _compute_rec(self, solution_set: np.ndarray) -> float:
...
         limited_solution_set = self._limit(point, solution_set)
         n_points_of_s = limited_solution_set.shape[0]
         if n_points_of_s == 1:
-            volume -= _compute_2points_volume(limited_solution_set[0], self._reference_point)
+            volume -= _compute_2points_volume(
+                limited_solution_set[0], np.array(self._reference_point)
```
Comment (hvy) [inline optuna/multi_objective/_hypervolume/wfg.py:58]: Ditto.

File: optuna/samplers/_tpe/multivariate_parzen_estimator.py [hunk 12]
```diff
@@ -265,18 +266,21 @@ def _transform_from_uniform(
             elif isinstance(distribution, distributions.DiscreteUniformDistribution):
                 q = self._q[param_name]
                 samples = np.round((samples - distribution.low) / q) * q + distribution.low
-                transformed[param_name] = np.clip(samples, distribution.low, distribution.high)
+                transformed[param_name] = np.array(
```
Comment (hvy) [inline optuna/samplers/_tpe/multivariate_parzen_estimator.py:269]:
```suggestion
                transformed[param_name] = np.asarray(
```

File: optuna/samplers/_tpe/multivariate_parzen_estimator.py [hunk 13]
```diff
@@ -265,18 +266,21 @@ def _transform_from_uniform(
...
+                    np.clip(samples, distribution.low, distribution.high)
+                )
             elif isinstance(distribution, distributions.IntUniformDistribution):
                 q = self._q[param_name]
+                assert q is not None
                 samples = np.round(samples / q) * q
-                transformed[param_name] = np.clip(
-                    samples, distribution.low, distribution.high
-                ).astype(int)
+                transformed[param_name] = np.array(
```
Comment (hvy) [inline optuna/samplers/_tpe/multivariate_parzen_estimator.py:276]:
```suggestion
                transformed[param_name] = np.asarray(
```

File: optuna/samplers/_tpe/multivariate_parzen_estimator.py [hunk 14]
```diff
@@ -265,18 +266,21 @@ def _transform_from_uniform(
...
+                    np.clip(samples, distribution.low, distribution.high)
+                )
             elif isinstance(distribution, distributions.IntLogUniformDistribution):
                 samples = np.round(np.exp(samples))
-                transformed[param_name] = np.clip(
-                    samples, distribution.low, distribution.high
-                ).astype(int)
+                transformed[param_name] = np.array(
```
Comment (hvy) [inline optuna/samplers/_tpe/multivariate_parzen_estimator.py:281]:
```suggestion
                transformed[param_name] = np.asarray(
```

File: optuna/samplers/_tpe/multivariate_parzen_estimator.py [hunk 15]
```diff
@@ -360,7 +366,7 @@ def _calculate_numerical_params(
             minsigma = 1.0 * (high - low) / min(100.0, (1.0 + len(mus)))
         else:
             minsigma = EPS
-        sigmas = np.clip(sigmas, minsigma, maxsigma)
+        sigmas = np.array(np.clip(sigmas, minsigma, maxsigma))
```
Comment (hvy) [inline optuna/samplers/_tpe/multivariate_parzen_estimator.py:369]:
```suggestion
        sigmas = np.asarray(np.clip(sigmas, minsigma, maxsigma))
```

**Review by hvy — APPROVED** [review]
Thanks @0x41head  for the overhauling PR. Looks great so let me approve this PR. (There a one comment left unaddressed but it's a minor https://github.com/optuna/optuna/pull/2300#discussion_r579953769)

**Comment by 0x41head** [comment]
Thank you @hvy , for helping me polish this PR.

File: optuna/samplers/_tpe/multi_objective_sampler.py [hunk 16]
```diff
@@ -464,8 +464,8 @@ def _sample_mo_categorical_index(
         above: np.ndarray,
     ) -> int:
         choices = distribution.choices
-        below = list(map(int, below))
-        above = list(map(int, above))
+        below = np.asarray(list(map(int, below)))
+        above = np.asarray(list(map(int, above)))
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/multi_objective_sampler.py:468]:
```suggestion
        below = below.astype(int)
        above = above.astype(int)
```

File: optuna/samplers/_tpe/multi_objective_sampler.py [hunk 17]
```diff
@@ -517,7 +517,7 @@ def _solve_hssp(
         hv_selected = 0.0
         while len(selected_indices) < subset_size:
             max_index = np.argmax(contributions)
-            contributions[max_index] = -1  # mark as selected
+            contributions[int(max_index)] = -1  # mark as selected
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/multi_objective_sampler.py:520]:
```python
max_index = int(np.argmax(contributions))
contributions[max_index] = -1  # mark as selected
```

File: optuna/samplers/_tpe/multivariate_parzen_estimator.py [hunk 18]
```diff
@@ -64,18 +64,19 @@ def __init__(
 
         self._mus: Dict[str, Optional[np.ndarray]] = {}
         self._sigmas: Dict[str, Optional[np.ndarray]] = {}
-        self._categorical_weights: Dict[str, Optional[np.ndarray]] = {}
+        self._categorical_weights: Dict[Optional[str], Optional[np.ndarray]] = {}
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/multivariate_parzen_estimator.py:67]: Can be reverted.
```suggestion
        self._categorical_weights: Dict[str, Optional[np.ndarray]] = {}
```

File: optuna/samplers/_tpe/multivariate_parzen_estimator.py [hunk 19]
```diff
@@ -64,18 +64,19 @@ def __init__(
...
         for param_name, dist in search_space.items():
             observations = multivariate_observations[param_name]
             if isinstance(dist, distributions.CategoricalDistribution):
                 mus = sigmas = None
-                categorical_weights = self._calculate_categorical_params(observations, param_name)
+                self._categorical_weights[param_name] = self._calculate_categorical_params(
+                    observations, param_name
+                )
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/multivariate_parzen_estimator.py:74]: Ditto.

File: optuna/samplers/_tpe/multivariate_parzen_estimator.py [hunk 20]
```diff
@@ -64,18 +64,19 @@ def __init__(
...
             else:
                 mus, sigmas = self._calculate_numerical_params(observations, param_name)
-                categorical_weights = None
+                self._categorical_weights[param_name] = None
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/multivariate_parzen_estimator.py:77]: Ditto.

File: optuna/samplers/_tpe/multivariate_parzen_estimator.py [hunk 21]
```diff
@@ -64,18 +64,19 @@ def __init__(
...
             self._mus[param_name] = mus
             self._sigmas[param_name] = sigmas
-            self._categorical_weights[param_name] = categorical_weights
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/multivariate_parzen_estimator.py:78]: Ditto.

File: optuna/samplers/_tpe/sampler.py [hunk 22]
```diff
@@ -324,14 +324,14 @@ def _split_observation_pairs(
         self, config_vals: List[Optional[float]], loss_vals: List[Tuple[float, float]]
     ) -> Tuple[np.ndarray, np.ndarray]:
 
-        config_vals = np.asarray(config_vals)
-        loss_vals = np.asarray(loss_vals, dtype=[("step", float), ("score", float)])
+        config_values = np.asarray(config_vals)
+        loss_values = np.asarray(loss_vals, dtype=[("step", float), ("score", float)])
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/sampler.py:328]: We can avoid introducing new variables by using `Union`. 
```suggestion
        config_vals: Union[List[Optional[float]], np.ndarray] = np.asarray(config_vals)
        loss_vals: Union[List[Optional[float]], np.ndarray] = np.asarray(loss_vals, dtype=[("step", float), ("score", float)])
```

File: optuna/samplers/_tpe/sampler.py [hunk 23]
```diff
@@ -341,17 +341,17 @@ def _split_multivariate_observation_pairs(
         loss_vals: List[Tuple[float, float]],
     ) -> Tuple[Dict[str, np.ndarray], Dict[str, np.ndarray]]:
 
-        config_vals = {k: np.asarray(v, dtype=float) for k, v in config_vals.items()}
-        loss_vals = np.asarray(loss_vals, dtype=[("step", float), ("score", float)])
+        config_values = {k: np.asarray(v, dtype=float) for k, v in config_vals.items()}
+        loss_values = np.asarray(loss_vals, dtype=[("step", float), ("score", float)])
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/sampler.py:345]: Ditto.

File: optuna/samplers/_tpe/sampler.py [hunk 24]
```diff
@@ -482,8 +482,8 @@ def _sample_categorical_index(
     ) -> int:
 
         choices = distribution.choices
-        below = list(map(int, below))
-        above = list(map(int, above))
+        blw = list(map(int, below))
+        ave = list(map(int, above))
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/sampler.py:486]:
```suggestion
        below = below.astype(int)
        above = above.astype(int)
```

File: tests/multi_objective_tests/samplers_tests/test_motpe.py [hunk 25]
```diff
@@ -585,7 +585,9 @@ def test_solve_hssp() -> None:
         truth = 0.0
         for subset in itertools.permutations(test_case, subset_size):
             truth = max(truth, sampler._compute_hypervolume(np.asarray(subset), r))
-        indices = sampler._solve_hssp(test_case, list(range(len(test_case))), subset_size, r)
+        indices = sampler._solve_hssp(
+            test_case, np.asarray(list(range(len(test_case)))), subset_size, r
+        )
```
Comment (HideakiImamura) [inline tests/multi_objective_tests/samplers_tests/test_motpe.py:590]: Can make it shorter
```suggestion
        indices = sampler._solve_hssp(test_case, np.arange(len(test_case)), subset_size, r)
```

File: tests/multi_objective_tests/samplers_tests/test_motpe.py [hunk 26]
```diff
@@ -599,7 +601,9 @@ def test_solve_hssp() -> None:
         truth = 0
         for subset in itertools.permutations(test_case, subset_size):
             truth = max(truth, sampler._compute_hypervolume(np.asarray(subset), r))
-        indices = sampler._solve_hssp(test_case, list(range(len(test_case))), subset_size, r)
+        indices = sampler._solve_hssp(
+            test_case, np.asarray(list(range(len(test_case)))), subset_size, r
+        )
```
Comment (HideakiImamura) [inline tests/multi_objective_tests/samplers_tests/test_motpe.py:606]: Ditto

File: tests/samplers_tests/tpe_tests/test_multi_objective_sampler.py [hunk 27]
```diff
@@ -618,7 +618,9 @@ def test_solve_hssp() -> None:
         truth = 0.0
         for subset in itertools.permutations(test_case, subset_size):
             truth = max(truth, sampler._compute_hypervolume(np.asarray(subset), r))
-        indices = sampler._solve_hssp(test_case, list(range(len(test_case))), subset_size, r)
+        indices = sampler._solve_hssp(
+            test_case, np.asarray(list(range(len(test_case)))), subset_size, r
+        )
```
Comment (HideakiImamura) [inline tests/samplers_tests/tpe_tests/test_multi_objective_sampler.py:623]: Ditto.

File: tests/samplers_tests/tpe_tests/test_multi_objective_sampler.py [hunk 28]
```diff
@@ -632,7 +634,9 @@ def test_solve_hssp() -> None:
         truth = 0
         for subset in itertools.permutations(test_case, subset_size):
             truth = max(truth, sampler._compute_hypervolume(np.asarray(subset), r))
-        indices = sampler._solve_hssp(test_case, list(range(len(test_case))), subset_size, r)
+        indices = sampler._solve_hssp(
+            test_case, np.asarray(list(range(len(test_case)))), subset_size, r
+        )
```
Comment (HideakiImamura) [inline tests/samplers_tests/tpe_tests/test_multi_objective_sampler.py:639]: Ditto.

**Review by HideakiImamura — CHANGES_REQUESTED** [review]
Sorry for the late review. I have several comments especially shadowing the name of variables. Please take a look?

**Comment by 0x41head** [comment]
Thanks for the review @HideakiImamura . I have committed the changes accordingly.

File: optuna/integration/botorch.py [hunk 29]
```diff
@@ -439,14 +440,16 @@ def sample_relative(
...
+        )
+        params: Union[numpy.ndarray, torch.Tensor]
+        con: Optional[Union[numpy.ndarray, torch.Tensor]] = None
+        bounds: Union[numpy.ndarray, torch.Tensor] = trans.bounds
         params = numpy.empty((n_trials, trans.bounds.shape[0]), dtype=numpy.float64)
-        con = None
-        bounds = trans.bounds
-
         for trial_idx, trial in enumerate(trials):
+
```
Comment (HideakiImamura) [inline optuna/integration/botorch.py:452]: Can be removed.
```suggestion
```

File: optuna/integration/botorch.py [hunk 30]
```diff
@@ -518,11 +521,9 @@ def sample_relative(
                 f"{candidates.size(0)}, bounds: {bounds.size(1)}."
             )
 
-        candidates = candidates.numpy()
-
-        params = trans.untransform(candidates)
+        parameters = trans.untransform(candidates.numpy())
 
-        return params
+        return parameters
```
Comment (HideakiImamura) [inline optuna/integration/botorch.py:526]: Can be one line.
```python
return trans.untransform(candidates.numpy())
```

File: optuna/samplers/_tpe/multivariate_parzen_estimator.py [hunk 31]
```diff
@@ -74,8 +75,8 @@ def __init__(
                 mus, sigmas = self._calculate_numerical_params(observations, param_name)
                 categorical_weights = None
             self._mus[param_name] = mus
-            self._sigmas[param_name] = sigmas
             self._categorical_weights[param_name] = categorical_weights
+            self._sigmas[param_name] = sigmas
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/multivariate_parzen_estimator.py:79]: Please move this line between L77 and L78. Looks like an unnecessary change.

File: optuna/samplers/_tpe/sampler.py [hunk 32]
```diff
@@ -324,14 +324,16 @@ def _split_observation_pairs(
...
-        config_vals = np.asarray(config_vals)
-        loss_vals = np.asarray(loss_vals, dtype=[("step", float), ("score", float)])
-
-        n_below = self._gamma(len(config_vals))
-        loss_ascending = np.argsort(loss_vals)
-        below = config_vals[np.sort(loss_ascending[:n_below])]
+        config_values: Union[List[Optional[float]], np.ndarray]
+        config_values = np.asarray(config_vals)
+        loss_values: Union[List[Optional[float]], np.ndarray]
+        loss_values = np.asarray(loss_vals, dtype=[("step", float), ("score", float)])
```
Comment (HideakiImamura) [inline optuna/samplers/_tpe/sampler.py:330]: Sorry for my previous inappropriate comment. Can be removed the type annotations. 
```suggestion
        config_values = np.asarray(config_vals)
        loss_values = np.asarray(loss_vals, dtype=[("step", float), ("score", float)])
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for your swift actions! I have other comments. PTAL.

**Comment by 0x41head** [comment]
Thanks for the review @HideakiImamura . PTAL

**Review by HideakiImamura — APPROVED** [review]
Thanks for the long run! LGTM!

**Comment by HideakiImamura** [comment]
Could you merge the master branch to pass the CI?

**Comment by 0x41head** [comment]
Merged the master optuna branch.

**Comment by HideakiImamura** [comment]
@0x41head Sorry for the late review. Could you merge the master branch again?

**Comment by 0x41head** [comment]
Manually resolved conflicts in `optuna/samplers/_tpe/multivariate_parzen_estimator.py` and merged the master branch.

**Review by HideakiImamura — APPROVED** [review]
Thanks for your swift actions! LGTM again!
